### PR TITLE
Fix ORCA returning wrong column type modifier info

### PR DESCRIPTION
**Pull request by hpbee**
**Issue:**
When executing a query with a false filter that results in zero rows 
from a table, ORCA returns incorrect column type information in the plan.
The type modifier details for columns in the table metadata are absent.

Setup query:
`CREATE TABLE foo(col decimal(7,4));`
Play query:
`EXPLAIN (VERBOSE) SELECT * FROM foo WHERE 1=0;`
```
                QUERY PLAN
------------------------------------------
 Result  (cost=0.00..0.00 rows=0 width=8)
   Output: NULL::numeric
   One-Time Filter: false
 Optimizer: GPORCA
(4 rows)
```
The type info `NULL::numeric` represents the column `col` of table `foo`
which is missing the modifier info of precision 7 and scale 4.

**RCA:**
In 7X, redundant Result node in a planned_stmt is being removed
by function `remove_redundant_results` in _orca.c_.
Within `remove_redundant_results`, the outer result node's var expr
is being replaced with child result's const expr.

The const expr representing a column(tle) in a child Result node 
of DXL or planned_stmt is created with default value of consttypmod.

So the replacement results in an erroneous consttypmod
for the column type information in the plan.

**Fix:**
To rectify this issue, the const.consttypmod value is now corrected 
before the replacement of var with const takes place.

Output from query above after fix:
```
                QUERY PLAN
------------------------------------------
 Result  (cost=0.00..0.00 rows=0 width=8)
   Output: NULL::numeric(7,4)
   One-Time Filter: false
 Optimizer: GPORCA
(4 rows)
```

**Regression test changes:**
* Added a new test case with the same query
 used by external libraries for fetching table schema.
* Fixed the expected output for a similar case.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/optimizer/plan/orca.c [hunk 1]
```diff
@@ -332,8 +332,16 @@ remove_redundant_results_mutator(Node *node, void *ctx)
...
-				tle->expr = (Expr *) push_down_expr_mutator((Node *) tle->expr,
-															child_plan->targetlist);
+				Expr *child_expr = (Expr *) push_down_expr_mutator((Node *) tle->expr, 
+																	child_plan->targetlist);
+				// When a const expr is created from a var,
+				// const.consttypmod is set as default value instead of var.vartypmod.
+				// const.consttypmod value needs to be fixed before replacing var with const.
+				if (IsA(tle->expr, Var) && IsA(child_expr, Const))
+				{
+					((Const *)child_expr)->consttypmod = ((Var *)tle->expr)->vartypmod;
```
Comment (gpopt) [inline src/backend/optimizer/plan/orca.c:342]: (Const *)child_expr, add a space after `)`

**Review by gpopt — APPROVED** [review]
The PR and commit title needs to be changed. It does not need to be related to Spark.

File: src/test/regress/expected/gporca_optimizer.out [hunk 2]
```diff
@@ -14874,3 +14874,20 @@ SELECT CAST(a AS TEXT[]) FROM array_coerceviaio;
...
+-- Below query is used by Java library to fetch schema of table.
+-- Test that the typmod of column type is correct in explain plan.
+EXPLAIN (VERBOSE, COSTS OFF) SELECT * FROM schema_test_table WHERE 1=0;
+                                                            QUERY PLAN                                                             
+-----------------------------------------------------------------------------------------------------------------------------------
+ Result
+   Output: NULL::numeric, NULL::numeric(5,2), NULL::character(10)
+   One-Time Filter: false
+ Optimizer: GPORCA
+ Settings: optimizer_enable_coordinator_only_queries = 'on', optimizer_enable_master_only_queries = 'on', optimizer_segments = '3'
```
Comment (gpopt) [inline src/test/regress/expected/gporca_optimizer.out:14890]: Can we remove the settings info here?

**Comment by chrishajas** [comment]
The code looks good, but the PR title/description should be updated with less focus on Spark. You're fixing an incorrect column type modifier that is returned by Orca (the "modifier" is important) that was not populated in certain cases (and you can describe the type of query: a const, false filter, etc)

**Comment by THANATOSLAVA** [comment]
> When a const expr is created from a var,
consttypmod is set as default value and it cannot replace a var as is,
which is happening in remove_redundant_results function.
Fixed the const.consttypmod value before replacing var with const.

We need to be more precise with the wording here. When you say "`Const` is **created** from `Var`", I think about casting, or some sort restructuring using the existing `Var`. But in this case, The `Const` expression comes from the `Result` node's child plan's target entry, and it's used to replace the `Result` node's target entry expression. What that means is we are tossing the `Var` structure, and putting a `Const` at its place.

Please update the PR description and comments accordingly.

File: src/backend/optimizer/plan/orca.c [hunk 3]
```diff
@@ -332,8 +332,16 @@ remove_redundant_results_mutator(Node *node, void *ctx)
...
+				Expr *child_expr = (Expr *) push_down_expr_mutator((Node *) tle->expr, 
+																	child_plan->targetlist);
+				// When a const expr is created from a var,
+				// const.consttypmod is set as default value instead of var.vartypmod.
+				// const.consttypmod value needs to be fixed before replacing var with const.
+				if (IsA(tle->expr, Var) && IsA(child_expr, Const))
+				{
+					((Const *) child_expr)->consttypmod = ((Var *) tle->expr)->vartypmod;
+				}
+				tle->expr = child_expr;
```
Comment (THANATOSLAVA) [inline src/backend/optimizer/plan/orca.c:344]: Can we move this logic in `push_down_expr_mutator`?

File: src/test/regress/sql/gporca.sql [hunk 4]
```diff
@@ -3608,6 +3608,15 @@ INSERT INTO array_coerceviaio values(ARRAY[1, 2, 3]);
 
 EXPLAIN SELECT CAST(a AS TEXT[]) FROM array_coerceviaio;
 SELECT CAST(a AS TEXT[]) FROM array_coerceviaio;
+
+---------------------------------------------------------------------------------
+DROP TABLE IF EXISTS schema_test_table;
+CREATE TABLE schema_test_table(a numeric, b numeric(5,2), c char(10) NOT NULL) distributed by (a);
+-- In 7x, redundant Result nodes in planned_stmt are being removed by ORCA,
+-- which caused the loss of typmod info of column type in plan.
+-- Below query is used by Java library to fetch schema of table.
```
[the remark this answers is not in the document]
Comment (hpbee) [reply src/test/regress/sql/gporca.sql:3617]: The information about why this query is useful in the test suite is being captured with this comment. Will remove that portion if that info is unnecessary.

File: src/backend/optimizer/plan/orca.c (hunk 3 above)
Comment (hpbee) [reply src/backend/optimizer/plan/orca.c:344]: Didn't want to do that in `push_down_expr_mutator` as that seemed like a more common function that can be used to fetch the replacement child expr of parent expr for other cases.
But on closer look it is used only in `remove_redundant_results_mutator`.
I will move the logic to `push_down_expr_mutator` as that func is used only for replacement of result right now.

**Comment by hpbee** [comment]
> The code looks good, but the PR title/description should be updated with less focus on Spark. You're fixing an incorrect column type modifier that is returned by Orca (the "modifier" is important) that was not populated in certain cases (and you can describe the type of query: a const, false filter, etc)

Made changes. Please take a look again.

File: src/backend/optimizer/plan/orca.c (hunk 3 above)
Comment (hpbee) [reply src/backend/optimizer/plan/orca.c]: Done. Please take a look.

**Comment by hpbee** [comment]
> > When a const expr is created from a var,
> > consttypmod is set as default value and it cannot replace a var as is,
> > which is happening in remove_redundant_results function.
> > Fixed the const.consttypmod value before replacing var with const.
> 
> We need to be more precise with the wording here. When you say "`Const` is **created** from `Var`", I think about casting, or some sort restructuring using the existing `Var`. But in this case, The `Const` expression comes from the `Result` node's child plan's target entry, and it's used to replace the `Result` node's target entry expression. What that means is we are tossing the `Var` structure, and putting a `Const` at its place.
> 
> Please update the PR description and comments accordingly.

@THANATOSLAVA Updated. Please take a look.

**Review by chrishajas — APPROVED** [review]

File: src/backend/optimizer/plan/orca.c [hunk 5]
```diff
@@ -332,8 +332,8 @@ remove_redundant_results_mutator(Node *node, void *ctx)
 			{
 				TargetEntry *tle = (TargetEntry *) lfirst(lc);
 
-				tle->expr = (Expr *) push_down_expr_mutator((Node *) tle->expr,
-															child_plan->targetlist);
+                                tle->expr = (Expr *) push_down_expr_mutator((Node *) tle->expr,
+																	child_plan->targetlist);
```
Comment (THANATOSLAVA) [inline src/backend/optimizer/plan/orca.c:336]: Same code. Please fix the indentation change.

File: contrib/postgres_fdw/expected/gp2pg_postgres_fdw_optimizer.out [hunk 6]
```diff
@@ -791,10 +791,10 @@ EXPLAIN (VERBOSE, COSTS OFF) SELECT * FROM ft1 t1 WHERE t1.c1 = 100 AND t1.c2 =
 (4 rows)
 
 EXPLAIN (VERBOSE, COSTS OFF) SELECT * FROM ft1 t1 WHERE c1 IS NULL;        -- NullTest
-                                                                                  QUERY PLAN                                                                                   
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+                                                                                        QUERY PLAN                                                                                        
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
```
Comment (THANATOSLAVA) [inline contrib/postgres_fdw/expected/gp2pg_postgres_fdw_optimizer.out:795]: Unnecessary diff

**Review by THANATOSLAVA — APPROVED** [review]

**Comment by THANATOSLAVA** [comment]
In your commit message,

**move**

"The const expr representing a column(tle) in a child Result node
of DXL or planned_stmt is created with default value of consttypmod."

**after**

"the outer result node's var expr
is being replaced with child result's const expr."

File: contrib/postgres_fdw/expected/gp2pg_postgres_fdw_optimizer.out (hunk 6 above)
Comment (hpbee) [reply contrib/postgres_fdw/expected/gp2pg_postgres_fdw_optimizer.out]: Fixed

File: src/backend/optimizer/plan/orca.c (hunk 5 above)
Comment (hpbee) [reply src/backend/optimizer/plan/orca.c]: Fixed

**Comment by hpbee** [comment]
Successful pipeline with latest changes: https://qpa.ci.gpdb.pivotal.io/teams/main/pipelines/gpdb-dev/jobs/compile_and_test_gpdb/builds/1802
